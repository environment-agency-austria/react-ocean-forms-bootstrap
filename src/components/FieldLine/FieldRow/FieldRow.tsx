/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { FormGroup } from 'reactstrap';

import { IFieldRowProps } from './FieldRow.types';

/**
 * Wrapper for FormGroup that generates additional
 * classes based on the meta state of the field
 */
export class FieldRow extends React.Component<IFieldRowProps> {
  public static displayName: string = 'FieldRow';

  public render(): JSX.Element {
    const {
      meta,
      children,
      className,
      ...rest
    } = this.props;

    const classes = className === '' || className === undefined ? [] : [className];

    if (!meta.valid) { classes.push('is-invalid'); }
    if (meta.touched) { classes.push('is-touched'); }

    const groupClass = classes.join(' ').trim();

    return (
      <FormGroup row className={groupClass} {...rest}>
        {children}
      </FormGroup>
    );
  }
}
