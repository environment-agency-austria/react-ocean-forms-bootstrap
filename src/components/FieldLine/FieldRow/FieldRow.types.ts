import { IFieldComponentMeta } from 'react-ocean-forms';

/**
 * Props for the FieldRow component
 */
export interface IFieldRowProps {
  /**
   * CSS class name to attach to the row
   */
  className?: string;
  /**
   * Field meta
   * @see IFieldComponentMeta
   */
  meta: IFieldComponentMeta;
}
