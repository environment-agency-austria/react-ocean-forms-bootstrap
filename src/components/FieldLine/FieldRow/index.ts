export * from './FieldRow';
export * from './FieldRow.types';
