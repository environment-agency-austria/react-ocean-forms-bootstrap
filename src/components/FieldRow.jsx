/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup } from 'reactstrap';
import { fieldMetaShape } from 'react-ocean-forms';

/**
 * Wrapper for FormGroup that generates additional
 * classes based on the meta state of the field
 */
function FieldRow({
  meta,
  children,
  className,
  ...rest
}) {
  const classes = [className];

  if (!meta.valid) classes.push('is-invalid');
  if (meta.touched) classes.push('is-touched');

  const groupClass = classes.join(' ').trim();

  return (
    <FormGroup row className={groupClass} {...rest}>
      {children}
    </FormGroup>
  );
}

FieldRow.defaultProps = {
  className: '',
};

FieldRow.propTypes = {
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  meta: fieldMetaShape.isRequired,
};

export default FieldRow;
