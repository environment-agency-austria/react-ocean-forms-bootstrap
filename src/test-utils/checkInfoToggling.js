/**
 * Re-usable test suite to check the info visibility
 * toggling between InfoAlert and InfoAddonButton
 */
const checkInfoToggling = (wrapper) => {
  describe('info visible toggling', () => {
    const getVisibility = () => wrapper.find('InfoAlert').prop('visible');

    const toggleCases = [
      ['InfoAddonButton', 'onClick'],
      ['InfoAlert', 'onClose'],
    ];

    describe.each(toggleCases)('%s click', (element, prop) => {
      let visibility = getVisibility();

      it(`should toggle the info visibility on ${element} click`, () => {
        for (let i = 0; i < 2; i += 1) {
          wrapper.find(element).prop(prop)();
          const newVisibility = getVisibility();
          expect(newVisibility).toBe(!visibility);
          visibility = newVisibility;
        }
      });
    });
  });
};

export default checkInfoToggling;
