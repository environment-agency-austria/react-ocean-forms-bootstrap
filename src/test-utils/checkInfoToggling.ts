import { ShallowWrapper } from 'enzyme';

/**
 * Re-usable test suite to check the info visibility
 * toggling between InfoAlert and InfoAddonButton
 */
export const checkInfoToggling = (wrapper: ShallowWrapper): void => {
  describe('info visible toggling', () => {
    const getVisibility = (): boolean => wrapper.find('InfoAlert').prop('visible');

    const toggleCases = [
      ['InfoAddonButton', 'onClick'],
      ['InfoAlert', 'onClose'],
    ];

    describe.each(toggleCases)('%s click', (element: string, prop: string) => {
      let visibility = getVisibility();

      it(`should toggle the info visibility on ${element} click`, () => {
        for (let i = 0; i < 2; i += 1) {
          const callback: Function = wrapper.find(element).prop(prop);
          callback();

          const newVisibility = getVisibility();
          expect(newVisibility).toBe(!visibility);
          visibility = newVisibility;
        }
      });
    });
  });
};
