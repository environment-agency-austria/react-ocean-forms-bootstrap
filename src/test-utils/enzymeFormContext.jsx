/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import createMockFormatter from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = registerCallback => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = () => ({
  valid: true,
  error: null,
  isValidating: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = () => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
