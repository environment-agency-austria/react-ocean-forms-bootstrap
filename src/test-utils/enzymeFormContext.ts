/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldState, IFormContext, TFormEventListener } from 'react-ocean-forms';

import { createMockFormatter } from './createMockFormatter';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = (registerCallback?: Function): IFormContext => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name: string, state: IFieldState): void => {
    if (registerCallback) {
      registerCallback(name, state);
    }
  }),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name: string, callback: TFormEventListener): void => {
    if (registerCallback) {
      registerCallback(name, callback);
    }
  }),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: createMockFormatter(),
  submit: jest.fn(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a mock field meta
 */
export const createMockFieldMeta = (): IFieldComponentMeta => ({
  valid: true,
  error: null,
  isValidating: false,
  isRequired: false,
  touched: false,
  stringFormatter: createMockFormatter(),
  plaintext: false,
});

/**
 * Creates a mock field shape
 */
export const createMockField = (): IFieldComponentFieldProps => ({
  id: 'mock-field',
  name: 'mock-field',
  value: 'mock-value',
  disabled: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
});
