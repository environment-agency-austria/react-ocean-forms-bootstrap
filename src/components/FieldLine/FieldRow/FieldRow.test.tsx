import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { IFieldComponentMeta } from 'react-ocean-forms';

import { createMockFieldMeta } from '../../../test-utils/enzymeFormContext';
import { FieldRow } from './FieldRow';
import { IFieldRowProps } from './FieldRow.types';

describe('<FieldRow />', () => {
  const baseClassName = 'mock-class';

  interface ISetupArgs {
    props?: Partial<IFieldRowProps>;
    metaOverrides?: Partial<IFieldComponentMeta>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
  }

  const setup = ({
    props,
    metaOverrides,
  }: ISetupArgs = {}): ISetupResult => {
    const meta = {
      ...createMockFieldMeta(),
      ...metaOverrides,
    };

    const wrapper = shallow((
      <FieldRow
        meta={meta}
        className={baseClassName}
        {...props}
      >
        <div id="mock" />
      </FieldRow>
    ));

    return {
      wrapper,
    };
  };

  it('should render correctly', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should work without the className prop', () => {
    const { wrapper } = setup({ props: { className: undefined } });
    expect(wrapper).toMatchSnapshot();
  });

  const metaCases = [
    ['field is invalid', 'valid', false, 'is-invalid'],
    ['field is touched', 'touched', true, 'is-touched'],
  ];

  describe.each(metaCases)('%s', (name: string, prop: string, value: boolean, className: string) => {
    it(`should set the class ${className} to the FormGroup if meta.${prop} is ${value}`, () => {
      const { wrapper } = setup({
        metaOverrides: { [prop]: value },
      });

      const groupClass = wrapper.find('FormGroup').prop('className');
      expect(groupClass).toEqual(expect.stringContaining(className));
      expect(groupClass).toEqual(expect.stringContaining(baseClassName));
    });

    it(`should NOT set the class ${className} to the FormGroup if meta.${prop} is ${!value}`, () => {
      const { wrapper } = setup({
        metaOverrides: { [prop]: !value },
      });

      const groupClass = wrapper.find('FormGroup').prop('className');
      expect(groupClass).toEqual(expect.not.stringContaining(className));
      expect(groupClass).toEqual(expect.stringContaining(baseClassName));
    });
  });
});
