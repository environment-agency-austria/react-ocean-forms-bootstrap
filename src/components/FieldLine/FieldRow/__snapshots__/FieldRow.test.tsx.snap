// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldRow /> should render correctly 1`] = `
<FormGroup
  className="mock-class"
  row={true}
  tag="div"
>
  <div
    id="mock"
  />
</FormGroup>
`;

exports[`<FieldRow /> should work without the className prop 1`] = `
<FormGroup
  className=""
  row={true}
  tag="div"
>
  <div
    id="mock"
  />
</FormGroup>
`;
